### ISSUE TITLE
Mezzio does not work without a router

### ISSUE THREAD

**Issue by boesing** [mezzio/mezzio#85]
### Bug Report

|    Q        |   A
|------------ | ------
| Version(s)  | 3.5.1

#### Summary

I have  use-case where I do not need routes but want to use mezzio.

The `Application` requires `RouteCollector` which is provided by the various router services. Due to the lack of an interface, the `RouteCollector` cannot be stripped away and thus, one has to provide a factory for the `RouteCollector` even tho the project does not need any routes.

#### Current behavior

When no router implementation is required by composer (and thus no factory is available for the `RouteCollector`), mezzio crashes.

```
Laminas\ServiceManager\Exception\ServiceNotCreatedException: Service with name "Mezzio\Router\RouteCollector" could not be created. Reason: Missing dependency "Mezzio\Router\RouterInterface" for service "Mezzio\Router\RouteCollector"; please make sure it is registered in your container. Refer to the Mezzio\Router\RouteCollector class and/or its factory to determine what the service should return.
```

#### How to reproduce

1. Create a new mezzio application using the `create-project` syntax as described within documentation
2. Remove the router you've chosen during setup with `composer remove mezzio/<router>`
3. Run application

#### Expected behavior

Until no route is provided to the application, nothing should happen.
When calling `$app->route` (or those aliases which wrap that method) on an app without a `RouteCollector`, an exception should be thrown.

**Comment by weierophinney** [comment]
@boesing I'm a bit confused - are you asking to allow NOT using a router, or that no errors are raised if the RouteCollector class is unavailable, but none of the methods that would use it are ever used?

**Comment by boesing** [comment]
@weierophinney In the first place, I would prefer having a `RouteCollectorInterface` which I can implement and override in my application to avoid having to require any of the implementing router components.
The `RouteCollector` is actually requiring the `RouterInterface` which is not available due to the lack of a router component.

I'm fine with that fact, but I'd still like to polyfill the `RouteCollectorInterface` with some kind of `ReadOnlyRouteCollector`.

I have a project which acts like some kind of "proxy". So *all* requests are passing middlewares and then are passed via another PSR-18 client to its real destination.
In this scenario, I dont have any routes as **any** requests are just piped through the application.
So there is neither a `RouteMiddleware` nor a `DispatchMiddleware` within my pipeline.

I currently just added the fastroute component to have a factory for the `RouterInterface` so that my application is bootable even tho I don't use that component at all.

This is why I've created that interface in mezzio/mezzio-router v3.5 and why I want to add it to `mezzio/mezzio` v3.6 in https://github.com/mezzio/mezzio/pull/87

Depending on the point of view, this is more a feature request than a bug report. :-/

**Comment by boesing** [comment]
Closed with #87
